### PR TITLE
Competing hazards recovery infection resolution

### PR DESCRIPTION
**Pull request by RJSheppard**
This PR resolves the competing hazards of infection and recovery. Giovanni wrote code to generate two competing hazards objects: CompetingOutcome (in which a function describing what happens to individuals following competing hazard resolution and the rates that that outcome occurs) and CompetingHazard (which links to each of the CompetingOutcome objects that require resolution, and which contains the process function that resolves the competing hazards), found in competing_hazards.R

We first replace create_biting_process and create_progression_process(es) with create_infection_rates_process and create_recovery_rate_process.

Create_infection_rate_process uses the renamed “biting_rate” (which was previously: create_biting_rate_process) which continues as previously, passing through the infection outcome object, and now ends when the rates of infection are generated. Note that we are currently only resolving infections in SA and U compartments, NOT D or Tr (consistent with deterministic model).  Note also that new infections are not rendered until after occurrence of resolution.

Similarly, all recovery/progression processes are now replaced with a single create_recovery_rates_process function, that uses calculate_recovery_rates, based on individual state rates of recovery.

The create_infection_recovery_hazard_process then uses the CompetingHazard object (to which each competing hazard outcome is assigned) to resolve competing hazards and determine which outcome occurs to which individuals.

The infection_outcome uses a new function: infection_process_resolved_hazard, which picks up from where we previously left off in the infection process, rendering incidence and assigning infections severities. Note that this function requires the infection rates to allow the incidence renderer to function.

The recovery_outcome then uses recovery_process_resolved_hazard, streamlining the progression functions into the relevant update functions. This function also passes in the dt variable for antimalarial resistance @tbreweric. (Hopefully it will still work!).

I ran simulations to compare the current default model (with mortality as the last process) with the resolved hazards. We do see something of an increase in incidence (which increases with EIR) and prevalence.
[competing_hazards_mortality_process_position.pdf](https://github.com/mrc-ide/malariasimulation/files/15190828/competing_hazards_mortality_process_position.pdf)
[competing_hazards_resolved_timesteps.pdf](https://github.com/mrc-ide/malariasimulation/files/15190830/competing_hazards_resolved_timesteps.pdf)

Human states also see fewer susceptibles, and is possibly deviating from the equilibrium a little.
[competing_hazards_resolved_states.pdf](https://github.com/mrc-ide/malariasimulation/files/15190885/competing_hazards_resolved_states.pdf)
[competing_hazards_resolved_states_time.pdf](https://github.com/mrc-ide/malariasimulation/files/15190895/competing_hazards_resolved_states_time.pdf)

Acquired immunities seem to be increasing a bit through time - consistent with higher onward infection following higher overall prevalence, particularly at lower EIR (higher plots).
[competing_hazards_resolved_immunity.pdf](https://github.com/mrc-ide/malariasimulation/files/15190880/competing_hazards_resolved_immunity.pdf)

I also checked through the expected infections and recoveries compared with the equilibrium solution, and found reasonable consistency, and resolution of the hazards. Let me know your thoughts!

### CODE REVIEW

File: R/disease_progression.R
Comment (giovannic) [inline R/disease_progression.R:16]: We probably want to cover all cases and use boolean instead of bitwise operators.

```suggestion
  if(isFALSE(parameters$antimalarial_resistance)){
    stopifnot(length(dt_input) == 1 && is.numeric(dt_input)) # please move this check to pre-simulation time for efficiency
    dt_v <- dt_input
  } else {
    dt_v <- dt_input$get_values(variables$state$get_index_of("Tr"))
  }
```

File: R/disease_progression.R
Comment (giovannic) [inline R/disease_progression.R:23]: It's perhaps more efficient to update variable called `recovery_rates` in the progression processes. Then all of this becomes.

```suggestion
  recovery_outcome$set_rates(variables$recovery_rates$get_values())
```

File: R/competing_hazards.R
Comment (giovannic) [inline R/competing_hazards.R:23]: We don't need this anymore? It's ugly 😆 

```suggestion
```

File: R/human_infection.R
Comment (giovannic) [inline R/human_infection.R:4]:
```suggestion
#' hazard resolution object and boosts immunity given infectious bites.
```

File: R/human_infection.R
Comment (giovannic) [inline R/human_infection.R:63]:
```suggestion
```

File: R/human_infection.R
Comment (giovannic) [inline R/human_infection.R:142]: Just a suggestion, this feels easier to understand

```suggestion
infection_outcome_process <- function(
```

File: R/human_infection.R
Comment (giovannic) [inline R/human_infection.R:526]: This whitespace actually helps to match brackets

```suggestion
    )
```

File: R/model.R
Comment (giovannic) [inline R/model.R:122]: Can we add this to `run_simulation` and `run_metapop_simulation` too?

File: R/processes.R
Comment (giovannic) [inline R/processes.R:71]:
```suggestion
      infection_outcome_process(timestep, target,
```

File: R/processes.R
Comment (giovannic) [inline R/processes.R:79]:
```suggestion
      recovery_outcome_process(timestep, target, variables, parameters, renderer)
```

File: R/processes.R
Comment (giovannic) [inline R/processes.R:95]: couldn't we keep the function factory in biting_process.R?

In this file we've been calling function factories from *_process.R files. But now we're making the functions in processes.R. It feels a bit inconsistent. And verbose since we have a lot of processes to create.

File: R/processes.R
Comment (giovannic) [inline R/processes.R:130]: Again, making the function here when it could be in disease_progression.R

File: R/human_infection.R
Comment (giovannic) [inline R/human_infection.R:145]: This is not used

```suggestion
```

**Review by giovannic — CHANGES_REQUESTED** [review]
This is a massive change. Thank you so much for the effort, it has been really thorough.

Few changes for readability and performance

**Comment by giovannic** [comment]
Also if you could rebase with dev, then we can run benchmarks to get an idea of the performance tradeoff.

**Review by pwinskill — COMMENTED** [review]
Thanks @RJSheppard. Some additional comments from me.

I was looking at the plots you submitted with the initial PR. Are these still valid now? The increase in prevalence for mid-EIRs was quite marked (if I was interpreting the plots correctly), which I think we should check and be cautious about.

I think, if those changes are that apparent we should probably look at PfPr~EIR plot similar to that in Penny et al. If possible lines for current and after changes would be helpful. 

image

File: R/render.R
Comment (giovannic) [reply R/render.R:192]: Yes! I thought I commented this, but that's gone missing. Good catch

File: R/render.R
Comment (RJSheppard) [reply R/render.R:113]: Whenever I look at this it always takes me a few to work out what it even means. I'd be happy to remove it, but agree it would be good to check if anyone uses it.

File: R/disease_progression.R
Comment (RJSheppard) [reply R/disease_progression.R:16]: I've made this change.

File: R/disease_progression.R
Comment (RJSheppard) [reply R/disease_progression.R:23]: I think this actually tidies up some of the issues I was having integrating with @tbreweric 's dt variable. I'm guessing that if we're storing all the recovery rates as a variable then we can replace the dt variable with a recovery_rates variable and still store all the relevant information. Let me know what you think with what I've done.

File: R/competing_hazards.R
Comment (RJSheppard) [reply R/competing_hazards.R:23]: Yep - gone!

File: R/human_infection.R
Comment (RJSheppard) [reply R/human_infection.R:4]: Made this change.

File: R/model.R
Comment (RJSheppard) [reply R/model.R:122]: So I think run_simulation calls the run_resumable_simulation anyway, so the incidence populate_incidence_rendering_columns should be called there. I've added a wrapper version that loops round the list of rendering objects to the run_metapop_simulation function which I think is working.

File: R/render.R
Comment (RJSheppard) [reply R/render.R:201]: Added this now - all seems to be working.

File: R/render.R
Comment (RJSheppard) [reply R/render.R:192]: I'm not sure I fully understand this. Is it as simple as moving the not repeating parts outside of the time looping section? The actual age rendering needs to occur at each time step, but the age rendering groups only need to be assigned once. I think I've done this now, but please let me know if it isn't what you meant!

**Comment by RJSheppard** [comment]
@pwinskill I'll rebase and make some plots as requested.

**Comment by RJSheppard** [comment]
[competing_hazards_inc_prev.pdf](https://github.com/user-attachments/files/16052882/competing_hazards_inc_prev.pdf)
[competing_hazards_states.pdf](https://github.com/user-attachments/files/16052883/competing_hazards_states.pdf)

Here are my latest comparison plots (@pwinskill) - as far as I can tell any difference has basically disappeared - possibly by increasing pop size and length of simulation.

**Comment by RJSheppard** [comment]
image
Immunity plots

**Review by giovannic — COMMENTED** [review]
Thanks! All my issues have been addressed.

Please remove the new dead code as commented.

Just going to wait for a benchmark run before passing the final review.

**Review by giovannic — APPROVED** [review]

**Review by plietar — APPROVED** [review]
I might try to get those missing 40% back later but it doesn't seem worth blocking merging this anymore.

**Comment by plietar** [comment]
All done from my end. There's not much more to be gained I think.

**Pull request merged**
